### PR TITLE
Backend SVG rendering proof of concept [ci skip]

### PR DESCRIPTION
**Pull request by camsaul**
Pursuant to #15776

Use a js bundle based on [airbnb's visx](https://airbnb.io/projects/visx/) to render charts in svg.

##### Workflow:

1. query results 
2. js charting library which creates svg string 
3. apache batik to render to png image 
4. put into html with hiccup
5. (optional if slack) screenshot hiccup with cssbox html rendering image

Some considerations we have to work around:
- batik svg->png cannot render html. So no chart output can include html (divs with text like legends), only svg
- cssbox does not render svgs, so we need those svg's to be rasterized and embedded in html

<details>
<summary>
Line Chart
</summary>

#### Before

image

#### After
image

</details>

<details>
<summary>
Bar chart
</summary>

#### Before
image
(yes that's not a bar chart but it should be)
#### After
image

</details>

<details>
<summary>
Pie chart
</summary>

#### Before
image

#### After
image

</details>

## More to do

there is more to do but we want to get this onto stats pronto.
Remaining:
- [ ] labels passed in rather than hardcoding
- [ ] don't assume everything is a date on the axis
- [ ] stretch goal(?) incorporate viz settings for colors
- [ ] multi-line graphs (we do custom axes on both sides, need to see if visx supports)

One weird thing that i would love thoughts on. Now, there is js code that is actually a dependency of the backend. This is built in the jar when building frontend deps, but now to run a dev server you need to build at least once `yarn build-static-viz`. I would like to put this into `clojure -T:deps prep` if that's possible or makes sense. Otherwise it's something else to remember that we need to do. Would love thoughts and ideas on this part.

### CODE REVIEW

File: dev/src/dev/render_png.clj [hunk 1]
```diff
@@ -2,13 +2,18 @@
   "Improve feedback loop for dealing with png rendering code"
   (:require [clojure.java.io :as io]
             [clojure.java.shell :as sh]
+            [clojure.string :as str]
+            [hiccup.core :as h]
             [metabase.models.card :as card]
             [metabase.models.user :as user]
             [metabase.pulse :as pulse]
             [metabase.pulse.render :as pulse-render]
+            [metabase.pulse.render.js-svg :as poc]
```
Comment (howonlee) [inline dev/src/dev/render_png.clj:11]: I feel like this is one of those theoretically-temporary acts of naming that will stay for an eyepopping amount of time if we don't think of another name besides `poc` right now. Maybe just "jssvg"? "js-render"?

File: dev/src/dev/render_png.clj [hunk 2]
```diff
@@ -2,13 +2,18 @@
...
+            [metabase.pulse.render.png :as png]
             [metabase.query-processor :as qp]
             [metabase.query-processor.middleware.permissions :as qp.perms]
-            [toucan.db :as tdb]))
+            [toucan.db :as tdb])
+  (:import org.fit.cssbox.misc.Base64Coder))
```
[the remark this answers is not in the document]
Comment (dpsutton) [reply dev/src/dev/render_png.clj:16]: I just grabbed the helper function to create a b64 encoded image byte string. Didn't really consider why it used a different one. This is all in a dev namespace to help people quickly see the output and limitations of the html rendering library.

File: dev/src/dev/render_png.clj [hunk 3]
```diff
@@ -58,3 +63,89 @@
...
+  [{:keys [chart html-file html-inline]}]
+  (let [chart-image (render-img-data-uri (svg-image chart))
+        chart-html (h/html [:img {:src chart-image :style "display: block; width: 100%"}])
+        html (cond html-file
+                   (slurp html-file)
+                   html-inline
+                   (str "<html><body style=\"margin: 0; padding: 0; background-color: white;\">"
+                        html-inline
+                        "</body></html>"))
+        html (h/html (str/replace html #"\{\{chart\}\}" chart-html))]
```
[the remark this answers is not in the document]
Comment (dpsutton) [reply dev/src/dev/render_png.clj:113]: True. This is just a simple dev helper with no more functionality expected.

File: src/metabase/pulse/render/js_engine.clj [hunk 4]
```diff
@@ -1,34 +1,39 @@
...
-   (doto (Context/newBuilder (u/varargs String ["js"]))
-     ;; these options allow JavaScript to access Java classes such as arrays directly -- see
-     ;; https://www.graalvm.org/reference-manual/js/JavaInteroperability/
-     (.allowHostAccess HostAccess/ALL)
-     (.allowHostClassLookup (reify java.util.function.Predicate
-                              (test [_ _] true))))))
+  (.. (Context/newBuilder (into-array String ["js"]))
+      (allowHostAccess HostAccess/ALL)
+      (allowHostClassLookup (reify java.util.function.Predicate
+                              (test [_ _] true)))
```
[the remark this answers is not in the document]
Comment (dpsutton) [reply src/metabase/pulse/render/js_engine.clj:11]: I'm gonna think on this. It's a good point. thanks.

File: src/metabase/pulse/render/js_svg.clj [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+(def ^:private ^Context context
+  "Javascript context suitable for evaluating the charts. It has the chart bundle and the above `src-api` in its
+  environment suitable for creating charts."
+  ;; todo is this thread safe? Should we have a resource pool on top of this? Or create them fresh for each invocation
+  (delay (load-viz-bundle (js/make-context))))
+
+(defn- parse-svg-string [^String s]
+  (let [s       (-> s
+                    (str/replace  #"<svg " "<svg xmlns=\"http://www.w3.org/2000/svg\" ")
+                    (str/replace #"fill=\"transparent\"" "fill-opacity=\"0.0\""))
```
[the remark this answers is not in the document]
Comment (dpsutton) [reply src/metabase/pulse/render/js_svg.clj]: I agree with you but unfortunately we cannot do that. The parser chokes if the svg tag does not have the xmln attribute and also on a fill=transparent attribute. I wish that was not the case but unfortunately it is.

```clojure
(parse-svg-string "<svg></svg>")
Execution error (IOException) at org.apache.batik.dom.util.SAXDocumentFactory/createDocument (SAXDocumentFactory.java:377).
Root element namespace does not match that requested:
Requested: http://www.w3.org/2000/svg
Found: null
```

It seems like the parser can handle the fill=transparent but the image renderer does not understand it. I've made a document tree transformer and pushed that. Thanks

File: dev/src/dev/render_png.clj (hunk 1 above)
Comment (dpsutton) [reply dev/src/dev/render_png.clj]: yeah i missed this rename. thanks

File: src/metabase/pulse/render/png.clj [hunk 6]
```diff
@@ -64,28 +65,28 @@
...
-          doc            (.parse (DefaultDOMSource. doc-source))
-          content-canvas (content-canvas doc doc-source width)]
-      (write-image! (.getImage content-canvas) "png" os))))
+  (with-open [is         (ByteArrayInputStream. (.getBytes html StandardCharsets/UTF_8))
+              doc-source (StreamDocumentSource. is nil "text/html; charset=utf-8")]
+    ;; todo: get rid of this hardcoded 1200. it is passed down from metabase.pulse.render/card-width = 400
+    (let [dimension       (Dimension. #_width 1200 1)
+          doc             (.parse (DefaultDOMSource. doc-source))
+          da              (dom-analyzer doc doc-source dimension)
+          graphics-engine (proxy [GraphicsEngine] [(.getRoot da) da (.getURL doc-source)]
```
[the remark this answers is not in the document]
Comment (dpsutton) [reply src/metabase/pulse/render/png.clj:77]: There's a call to `.createLayout` and `.getImage` below. The graphics engine here is supposed to be subclassed to add rendering hints which is what this code is doing. There's no good way to just pass in the rendering hints as it is very OOP style. I was trying to write a comment but they all largely mimic exactly what the code is doing.

"Proxy so we can override the call to `setupGraphics` to add rendering hints for quality" But the code is essentially exactly that. It is a proxy for `GraphicsEngine` and one method `setupGraphics` is overridden to add rendering hints. So i'm not sure it adds much.

File: src/metabase/pulse/render/png.clj [hunk 7]
```diff
@@ -64,28 +65,28 @@
...
   (register-fonts-if-needed!)
-  (with-open [is (ByteArrayInputStream. (.getBytes html StandardCharsets/UTF_8))]
-    (let [doc-source     (StreamDocumentSource. is nil "text/html; charset=utf-8")
-          doc            (.parse (DefaultDOMSource. doc-source))
-          content-canvas (content-canvas doc doc-source width)]
-      (write-image! (.getImage content-canvas) "png" os))))
+  (with-open [is         (ByteArrayInputStream. (.getBytes html StandardCharsets/UTF_8))
+              doc-source (StreamDocumentSource. is nil "text/html; charset=utf-8")]
+    ;; todo: get rid of this hardcoded 1200. it is passed down from metabase.pulse.render/card-width = 400
+    (let [dimension       (Dimension. #_width 1200 1)
```
[the remark this answers is not in the document]
Comment (dpsutton) [reply src/metabase/pulse/render/png.clj:74]: Good catch. I need to figure this bit out

**Review by jeff303 — APPROVED** [review]

**Pull request merged**